### PR TITLE
Update to arcade 6.0

### LINKED ISSUE #6119: Upgrade To Arcade 6.0
This is to get MSBuild on [arcade powered source-build](https://github.com/dotnet/source-build/tree/master/Documentation/planning/arcade-powered-source-build).

/cc @dagood

### PR DESCRIPTION
**Pull request by BenVillalobos**
Fixes #6119

### Context
This is to get MSBuild on [arcade powered source-build](https://github.com/dotnet/source-build/tree/master/Documentation/planning/arcade-powered-source-build).

### Changes Made
Ran `darc update-dependencies --source-repo arcade --channel ".NET 5 Eng - Latest"`

3/3: Ran `darc update-dependencies --source-repo arcade --channel ".NET Eng - Latest"`

### Testing
CI is enough.

### Notes
I expect this PR to be in flight as I chip away at it over time. Or it will magically work first time through.

For reviewers: The first commit is the biggest diff and was entirely generated by running the darc command above, so you may want to start at commit 2 when it comes up.

### CODE REVIEW

**Comment by BenVillalobos** [comment]
PR looking good so far.

Should look into this before merging. May just need to update a path to look for 3.1.
image

**Comment by Forgind** [comment]
I thought the whole reason we didn't go straight to 6.0 was that we didn't want preview bits in release MSBuild?

**Comment by BenVillalobos** [comment]
@Forgind that's at odds with what source-build needs, which is for us to be on arcade 6. @marcpopMSFT thoughts? arcade 6 is LTS FYI.

**Comment by BenVillalobos** [comment]
VS PR: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/303587

**Comment by BenVillalobos** [comment]
The VS PR failed almost identically to https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/303587, which suggests that merging arcade 6 won't cause any issues (famous last words).

We should wait until master is unblocked before merging this.

**Comment by BenVillalobos** [comment]
@epinanth, is there a channel for "I want to be on and stay on arcade 6" yet?

**Comment by BenVillalobos** [comment]
@epananth friendly ping (mispelled your username in my previous post). Seeing a lot of strange errors running `darc update-dependencies --source-repo arcade --channel ".NET Eng - Latest"`. Should we not be using this channel if we want to be on arcade 6?

**Comment by epananth** [comment]
Going to look at this next week, I'm on FR next week

**Comment by BenVillalobos** [comment]
@epananth friendly ping 🙂

**Comment by epananth** [comment]
Looking into this one today, I will post an update soon.

**Comment by BenVillalobos** [comment]
Notes to self: I was wondering why this succeeded on my local machine but was seeing "sdk not found" issues for CI builds. Turns out I had the net5 sdk installed under the .dotnet folder in the repo. Time to start updating projects to net6.0

**Comment by BenVillalobos** [comment]
We're hitting this breaking change by updating to net6.0: https://github.com/dotnet/docs/issues/19625

Thread.ResetAbort() is obsolete: https://source.dot.net/#System.Private.CoreLib/Thread.cs,47d38cd18e02048c

Looks like we can just ifdef around this.

**Comment by BenVillalobos** [comment]
I can't seem to repro the (no bootstrap) build. Even after running `git clean -xdf` and removing my 6.0.100 preview sdk.

Got it to repro on linux and seeing the same "could not find nuget.frameworks.dll" error. Not sure what the root cause is.

Found this when digging through the assembly
```
// NuGet.Frameworks, Version=5.10.0.7148, Culture=neutral, PublicKeyToken=31bf3856ad364e35
	// Assembly reference loading information:
	// There were some problems during assembly reference load, see below for more information!
	// Error: Could not find reference: NuGet.Frameworks, Version=5.10.0.7148, Culture=neutral, PublicKeyToken=31bf3856ad364e35
```

`ProcessFrameworkReferences` is an sdk task, any ideas @dsplaisted?
image

**Comment by Forgind** [comment]
Try merging in master?

**Comment by BenVillalobos** [comment]
@Forgind Not sure what you mean? latest/master into this branch? 

It was updated to latest main recently.

**Comment by BenVillalobos** [comment]
It's worth a shot!

**Comment by BenVillalobos** [comment]
Tried building from a VM with a clean VS install and running `eng\cibuild_bootstrapped_msbuild.cmd` and no repro.

File: eng/Version.Details.xml
Comment (rainersigwald) [inline eng/Version.Details.xml:18]: Self, is this actually a _product_ dependency?

File: eng/Versions.props
Comment (rainersigwald) [inline eng/Versions.props:34]: Shouldn't darc have been doing this for us? Are we missing a subscription?

File: ref/Microsoft.Build/Microsoft.Build.cs
Comment (rainersigwald) [inline ref/Microsoft.Build/Microsoft.Build.cs:1]: Something's up here; this shouldn't be a new file.

File: src/Directory.BeforeCommon.targets
Comment (rainersigwald) [inline src/Directory.BeforeCommon.targets:125]: IIRC the SDK defines something much like this. Can we use that instead?

Comment (BenVillalobos) [reply src/Directory.BeforeCommon.targets:125]: running `rg -Fi "<defineconstants" -tmsbuild` in the sdk doesn't look promising. Sounds like something we can merge into the SDK to allow compiler directives in code for breaking changes between sdk versions. Something like:
```xml
<DefineConstants Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'netcoreapp3.1'))">$(DefineConstants);NET31_PLUS</DefineConstants>
<DefineConstants Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net5.0'))">$(DefineConstants);NET5_PLUS</DefineConstants>
<DefineConstants Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">$(DefineConstants);NET6_PLUS</DefineConstants>
```
@dsplaisted does this sound reasonable?

Questions answered with some greppin'
1. Where is DefineConstants even consumed?
   - Looks like it's passed as an argument to the compiler. https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.CSharp.CurrentVersion.targets#L249

File: eng/Versions.props
Comment (BenVillalobos) [reply eng/Versions.props:34]: You're right: https://github.com/dotnet/msbuild/pull/6327 looks to be the last time this version updated automatically. 

I'm thinking it was either the master->main rename or the switch from nuget.client.trusted to nuget.client that broke this.

Comment (BenVillalobos) [reply eng/Versions.props:34]: Just going to log my thought process here.

Likely not due to the switch from nuget.client.trusted since we got a few PR's since then. It's more likely that it's darc.

Comment (BenVillalobos) [reply eng/Versions.props:34]: Looks like there's no sub for it.
`darc get-subscriptions --channel "VS 16.10"`
```
https://github.com/dotnet/fsharp (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('main')
  - Id: 1b3ed604-a240-43c5-f3e6-08d8e97c775d
  - Update Frequency: EveryDay
  - Enabled: True
  - Batchable: False
  - Merge Policies: []
  - Last Build: 20210427.5 (b6a352ab0a11a0584348dd77811af7eecaf9d072)
https://github.com/dotnet/fsharp (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/5.0.3xx')
  - Id: 8d76063a-5130-4369-8278-08d8e9754760
  - Update Frequency: EveryDay
  - Enabled: True
  - Batchable: False
  - Merge Policies:
    Standard
  - Last Build: 20210430.7 (e411df7595bbe379f07dd2bcbb6e4046eca9eb31)
https://github.com/dotnet/fsharp (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/6.0.1xx-preview4')
  - Id: fec82357-f505-450c-c78c-08d8ff796cf5
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: False
  - Merge Policies:
    Standard
  - Last Build: 20210430.7 (e411df7595bbe379f07dd2bcbb6e4046eca9eb31)
https://github.com/dotnet/msbuild (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('main')
  - Id: 9b9c2fc1-dee5-447e-827b-08d8e9754760
  - Update Frequency: EveryDay
  - Enabled: True
  - Batchable: False
  - Merge Policies:
    Standard
  - Last Build: 20210428.2 (5b9216a75e98e19eba84e04a5f30bd35a68f317a)
https://github.com/dotnet/msbuild (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/5.0.3xx')
  - Id: 08d8569c-a921-4400-1fc7-08d8c8766580
  - Update Frequency: EveryDay
  - Enabled: True
  - Batchable: False
  - Merge Policies:
    Standard
  - Last Build: 20210428.2 (5b9216a75e98e19eba84e04a5f30bd35a68f317a)
https://github.com/dotnet/msbuild (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/6.0.1xx-preview4')
  - Id: f3da650c-3ec0-4a87-5a84-08d8ff79a3bc
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: False
  - Merge Policies:
    Standard
  - Last Build: 20210503.2 (fa96a2a81e0fb8c028057fa204bbf386bfb36aec)
https://github.com/dotnet/roslyn (VS 16.10) ==> 'https://github.com/dotnet/arcade' ('main')
  - Id: 71f51b42-1554-444f-7918-08d8e262a02c
  - Update Frequency: EveryWeek
  - Enabled: True
  - Batchable: True
  - Merge Policies:
    Standard
  - Last Build: 20210501.8 (4c32f5e4e9c0828a94fd4d1c9c0994082c85aaf3)
https://github.com/microsoft/vstest (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/5.0.3xx')
  - Id: b4754456-bab5-48e6-093c-08d8ff7930ba
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: False
  - Merge Policies: []
  - Last Build: 20210429-01 (1050e72d7a92a5fb5737aa6a464e8e43e7dc52f4)
https://github.com/nuget/nuget.client (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('main')
  - Id: fcb199f6-ff33-44a0-f3ef-08d8e97c775d
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: False
  - Merge Policies: []
  - Last Build: 5.10.0.7228 (d2bb0de35242b04802c1a7856ebb7718f674565c)
https://github.com/nuget/nuget.client (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/5.0.3xx')
  - Id: 61b07feb-68a2-4b3c-64a4-08d8e9757c68
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: False
  - Merge Policies: []
  - Last Build: 5.10.0.7228 (d2bb0de35242b04802c1a7856ebb7718f674565c)
https://github.com/nuget/nuget.client (VS 16.10) ==> 'https://github.com/dotnet/sdk' ('release/6.0.1xx-preview4')
  - Id: 5184e9dc-a7d8-43a0-c78a-08d8ff796cf5
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: False
  - Merge Policies:
    Standard
  - Last Build: 5.10.0.7240 (dca1d060f38e1e02f6bfca41e25f081f19fd534b)
```

Comment (BenVillalobos) [reply eng/Versions.props:34]: Correctomundo! `darc add-subscription --channel "VS 16.10" --source-repo "https://github.com/nuget/nuget.client" --target-repo "https://github.com/dotnet/msbuild" --target-branch main --update-frequency "everyWeek" --trigger`

See https://github.com/dotnet/msbuild/pull/6405

File: src/Directory.BeforeCommon.targets
Comment (rainersigwald) [reply src/Directory.BeforeCommon.targets:125]: https://github.com/dotnet/sdk/pull/14798 added it; in this case it's `NET6_0_OR_GREATER`.

**Review by rainersigwald — APPROVED** [review]
LGTM pending #6408.

File: eng/Version.Details.xml
Comment (Forgind) [inline eng/Version.Details.xml:12]: Do we own this/should we be specifying its version? Do we have a plan for keeping it up-to-date with the version specified in Versions.props?

Comment (rainersigwald) [reply eng/Version.Details.xml:12]: yes: darc subscription. Can't set it up until this merges though.

**Review by Forgind — APPROVED** [review]

**Comment by MichaelSimons** [comment]
Can this get merged now?  The one issue I saw mentioned as blocking appears to have been merged - #6408.  This is blocking https://github.com/dotnet/msbuild/pull/6387.  Thanks

**Pull request merged**
